File: rb/utility/prepare_auth.rb
# XenoCanto SDK utility: prepare_auth
require_relative 'struct/voxgig_struct'
module XenoCantoUtilities
  QUERY_AUTH = "key"
  OPTION_APIKEY = "apikey"
  NOT_FOUND = "__NOTFOUND__"

  PrepareAuth = ->(ctx) {
    spec = ctx.spec
    return nil, ctx.make_error("auth_no_spec", "Expected context spec property to be defined.") unless spec

    query = spec.query
    options = ctx.client.options_map

    # Public APIs that need no auth omit the options.auth block entirely.
    if options["auth"].nil?
      query.delete(QUERY_AUTH)
      return spec, nil
    end

    apikey = VoxgigStruct.getprop(options, OPTION_APIKEY, NOT_FOUND)

    if apikey.nil? || (apikey.is_a?(String) && (apikey == NOT_FOUND || apikey == ""))
      query.delete(QUERY_AUTH)
    else
      apikey_val = apikey.is_a?(String) ? apikey : ""
      # NO PREFIX IN A QUERY STRING: `?token=Bearer%20abc` is not a thing
      # any API reads, so options.auth.prefix is dropped rather than joined.
      query[QUERY_AUTH] = apikey_val
    end

    return spec, nil
  }
end

File: go/utility/prepare_auth.go
package utility

import (
	vs "github.com/voxgig-sdk/xeno-canto-sdk/go/utility/struct"

	"github.com/voxgig-sdk/xeno-canto-sdk/go/core"
)

const credName = "key"
const optionApikey = "apikey"
const notFound = "__NOTFOUND__"

func prepareAuthUtil(ctx *core.Context) (*core.Spec, error) {
	spec := ctx.Spec
	if spec == nil {
		return nil, ctx.MakeError("auth_no_spec",
			"Expected context spec property to be defined.")
	}

	query := spec.Query
	options := ctx.Client.OptionsMap()

	// Public APIs that need no auth omit the options.auth block entirely.
	if options["auth"] == nil {
		delete(query, credName)
		return spec, nil
	}

	apikey := vs.GetProp(options, optionApikey, notFound)

	skip := false
	if apikey == nil {
		skip = true
	} else if apikeyStr, ok := apikey.(string); ok &&
		(apikeyStr == notFound || apikeyStr == "") {
		skip = true
	}

	if skip {
		delete(query, credName)
	} else {
		apikeyVal := ""
		if av, ok := apikey.(string); ok {
			apikeyVal = av
		}
		query[credName] = apikeyVal
	}

	return spec, nil
}
